File: Chapter 7-UFOs/Classes/UFOViewController/UFOViewController.m
//
//  GameCenterViewController.m
//  GameCenter
//
//  Created by Kyle Richter on 11/14/10.
//  Copyright 2010 Dragon Forged Software. All rights reserved.
//

#import "UFOViewController.h"
#import "UFOGameViewController.h"
#import "UFOLeaderboardViewController.h"
#import "UFOAchievementViewController.h"

@implementation UFOViewController

-(void)viewDidLoad
{
	[super viewDidLoad];
	
	if([GameCenterManager isGameCenterAvailable])
	{
		[[NSNotificationCenter defaultCenter] addObserver:self
												 selector:@selector(localUserAuthenticationChanged:)
													 name:GKPlayerAuthenticationDidChangeNotificationName 
												   object:nil];
		
		gcManager = [[GameCenterManager alloc] init];
		gcManager.delegate = self;
        [gcManager authenticateLocalUser: self];
		[gcManager populateAchievementCache];
		//[gcManager resetAchievements];
		[gcManager findAllActivity];
		
	}
	
	else
		NSLog(@"Game Center not available");	
    
}

-(void)viewDidUnload
{
	[[NSNotificationCenter defaultCenter] removeObserver:self name:GKPlayerAuthenticationDidChangeNotificationName object:nil];	
	
	[super viewDidUnload];
}	

-(void)viewWillAppear:(BOOL)animated
{
	gcManager.delegate = self;
	[super viewWillAppear:animated];
}

- (void)playerDataLoaded:(NSArray *)players error:(NSError *)error;
{
	if(error != nil)
		NSLog(@"An error occured during player lookup: %@", [error localizedDescription]);
	else 
		NSLog(@"Players loaded: %@", players);
}

-(void)localUserAuthenticationChanged:(NSNotification *)notif; 
{
	NSLog(@"Authenication Changed: %@", notif.object);	
}

- (void)processGameCenterAuthentication:(NSError*)error;
{
	if(error != nil)
	{
		NSLog(@"An error occured during authentication: %@", [error localizedDescription]);
	}
	else 
	{
		[gcManager setupInvitationHandler:self];
	}
}

- (void)findProgrammaticMatch
{
	GKMatchRequest *request = [[[GKMatchRequest alloc] init] autorelease]; 
	request.minPlayers = 2;
	request.maxPlayers = 4;
	
    
	[[GKMatchmaker sharedMatchmaker] findMatchForRequest:request withCompletionHandler:^(GKMatch *match, NSError *error) {
		if (error) 
		{
			NSLog(@"An error occurrred during finding a match: %@", [error localizedDescription]);
		} 
		
		else if (match != nil)
		{
			NSLog(@"A match has been found: %@", match);
		}
	}];
}

- (void)findProgrammaticHostedMatch
{
	GKMatchRequest *request = [[[GKMatchRequest alloc] init] autorelease]; 
	request.minPlayers = 2;
	request.maxPlayers = 16;
    [[GKMatchmaker sharedMatchmaker] findPlayersForHostedRequest:request withCompletionHandler:^(NSArray<GKPlayer *> * _Nullable players, NSError * _Nullable error) {
        if (error) {
            NSLog(@"An error occurrred during finding a match: %@", [error localizedDescription]);
        } else if (players != nil) {
            NSLog(@"Players have been found for match: %@", players);
        }
    }];
}

- (void)addPlayerToMatch:(GKMatch *)match withRequest:(GKMatchRequest *)request
{
	[[GKMatchmaker sharedMatchmaker] addPlayersToMatch:match matchRequest:request completionHandler:^(NSError *error)
	{
		if (error) 
		{
			NSLog(@"An error occurrred during adding a player to match: %@", [error localizedDescription]);
		} 
		
		else if (match != nil)
		{
			NSLog(@"A player has been added to the match");
		}
	}];
}


- (void)friendsFinishedLoading:(NSArray *)friends error:(NSError *)error;
{
	if(error != nil)
		NSLog(@"An error occured during friends list request: %@", [error localizedDescription]);
	else 
		[gcManager playersForIDs: friends];
}

- (void)scoreReported: (NSError*) error;
{
	if(error)
		NSLog(@"There was an error in reporting the score: %@", [error localizedDescription]);
	
	else	
		NSLog(@"Score submitted");
}


- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
	if(interfaceOrientation == UIInterfaceOrientationLandscapeLeft || interfaceOrientation == UIInterfaceOrientationLandscapeRight)
		return YES;
	
	return NO;
}

-(IBAction)playButtonPressed
{
	UFOGameViewController *gameViewController = [[UFOGameViewController alloc] init];
	gameViewController.gcManager = gcManager;
	[self.navigationController pushViewController: gameViewController animated:YES];
	[gameViewController release];
}

-(IBAction)leaderboardButtonPressed;
{
    GKGameCenterViewController *leaderboardController = [[GKGameCenterViewController alloc] initWithState: leaderboardController];
    leaderboardController.gameCenterDelegate = self;
    [self presentViewController:leaderboardController animated:YES completion:nil];
}

-(IBAction)customLeaderboardButtonPressed;
{
	UFOLeaderboardViewController *leaderboardViewController = [[UFOLeaderboardViewController alloc] init];
	leaderboardViewController.gcManager = gcManager;
    [self presentViewController:leaderboardViewController animated:YES completion:nil];
	[leaderboardViewController release];
}

- (void)gameCenterViewControllerDidFinish:(GKGameCenterViewController *)gameCenterViewController {
    [self dismissViewControllerAnimated:YES completion: nil];
}

-(IBAction)achievementButtonPressed;
{
    GKGameCenterViewController *achievementViewController = [[GKGameCenterViewController alloc] initWithState: 1];
    achievementViewController.gameCenterDelegate = self;
    [self presentViewController:achievementViewController animated:YES completion:nil];
}

-(IBAction)customAchievementButtonPressed;
{
	UFOAchievementViewController *achievementViewController = [[UFOAchievementViewController alloc] init];
	achievementViewController.gcManager = gcManager;
    [self presentViewController:achievementViewController animated:YES completion:nil];
	[achievementViewController release];
}

-(IBAction)multiplayerButtonPressed;
{
	GKMatchRequest *request = [[[GKMatchRequest alloc] init] autorelease];
	request.minPlayers = 2; 
	request.maxPlayers = 4;
	
	GKMatchmakerViewController *matchmakerViewController = [[[GKMatchmakerViewController alloc] initWithMatchRequest:request] autorelease];
	matchmakerViewController.matchmakerDelegate = self; 
		
    [self presentViewController:matchmakerViewController animated:YES completion:nil];
}

- (void)matchmakerViewControllerWasCancelled:(GKMatchmakerViewController *)viewController 
{
    [self dismissViewControllerAnimated:YES completion:nil];
}

//MARK: Depreacated Function
- (void)matchmakerViewController:(GKMatchmakerViewController *)viewController didFailWithError:(NSError *)error 
{
    [self dismissViewControllerAnimated:YES completion:nil];
	
	if(error != nil)
	{
        UIAlertController *alert = [UIAlertController
                                    alertControllerWIthTitle:@""
                                    message:[NSString stringWithFormat:@"An error occurred: %@", [error localizedDescription]]
                                    ];
//											  delegate:nil
//											  cancelButtonTitle:@"Dismiss"
//											  otherButtonTitles:nil];
        UIAlertAction *defaultAction = [UIAlertAction
                                    actionWithTitle:@"Dismiss" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {}];
    
        [alert addAction:defaultAction];
        
        [self presentViewController:alert animated:YES completion:nil];
	}	
}

- (void)matchmakerViewController:(GKMatchmakerViewController *)viewController didFindMatch:(GKMatch *)match 
{
    [self dismissViewControllerAnimated:YES completion:nil];
    
    gcManager.matchOrSession = match;
    
    [gcManager setupInvitationHandler: gcManager];

    UFOGameViewController *gameVC = [[UFOGameViewController alloc] init];
    gameVC.gcManager = gcManager;
    gameVC.gameIsMultiplayer = YES;
    gameVC.peerIDString = nil;
    gameVC.peerMatch = match;
    [self.navigationController pushViewController:gameVC animated:YES];
    [gameVC release];
}

- (void)matchmakerViewController:(GKMatchmakerViewController *)viewController didFindPlayers:(NSArray *)playerIDs
{
    [self dismissViewControllerAnimated:YES completion:nil];
	NSLog(@"Players: %@", playerIDs);
	//Begin Hosted Game
}

- (void)playerActivity:(NSNumber *)activity error:(NSError *)error
{
	if(error != nil)
	{
		NSLog(@"An error occurred while querying player activity: %@", [error localizedDescription]);
	}	
	
	else
	{
		NSLog(@"All recent player activity: %@", activity);
	}
}

- (void)playerActivityForGroup:(NSDictionary *)activityDict error:(NSError *)error
{
	if(error != nil)
	{
		NSLog(@"An error occurred while querying player activity: %@", [error localizedDescription]);
	}	
	
	else
	{
		NSLog(@"All recent player activity: %@ For Group: %@", [activityDict objectForKey:@"activity"], [activityDict objectForKey:@"group"]);
	}
}	


-(IBAction)localMultiplayerGameButtonPressed
{
	peerPickerController = [[MCBrowserViewController alloc] initWithServiceType:@"space" session:currentSession];
	peerPickerController.delegate = self;
//	peerPickerController.connectionTypesMask = GKPeerPickerConnectionTypeNearby;
    [peerPickerController presentViewController:peerPickerController animated:true completion:nil];
}


- (void)setupSession {
    MCPeerID *peer = [[MCPeerID alloc] initWithDisplayName: [[GKLocalPlayer localPlayer] alias]];
    currentSession = [[MCSession alloc] initWithPeer:peer];
    currentSession.delegate = gcManager;
    advertiserAssistant = [[MCAdvertiserAssistant alloc] initWithServiceType:@"space" discoveryInfo:nil session: currentSession];
    [advertiserAssistant start];
}

//- (void)peerPickerController:(MCBrowserViewController *)picker didSelectConnectionType:(MCNearbyServiceBrowser *)type
//{
//	if(type == GKPeerPickerConnectionTypeOnline)
//	{
//        [picker dismissViewControllerAnimated:true completion:nil];
//        [picker release];
//
//		// Display your own user interface here.
//	}
//}

- (void)peerPickerController:(MCBrowserViewController *)picker
              didConnectPeer:(NSString *)peerID
                   toSession:(MCSession *)session
{
    [picker dismissViewControllerAnimated:true completion:nil];

	currentSession = session;
    gcManager.matchOrSession = session;

    [session setDataReceiveHandler: gcManager  withContext: nil];

    UFOGameViewController *gameVC = [[UFOGameViewController alloc] init];
    gameVC.gcManager = gcManager;
    gameVC.gameIsMultiplayer = YES;
    gameVC.peerIDString = peerID;
    gameVC.peerMatch = nil;
    [self.navigationController pushViewController:gameVC animated:YES];
    [gameVC release];
}

-(IBAction)storeButtonPressed;
{
    
    UFOStoreViewController *storeVC = [[UFOStoreViewController alloc] init];
    [self presentViewController:storeVC animated:TRUE completion:nil];
}

@end

File: Chapter 5-UFOs/Classes/UFOViewController/UFOViewController.swift
//
//  UFOViewController.swift
//  GameCenter
//
//  Created by Kyle Richter on 11/14/10.
//  Copyright 2010 Dragon Forged Software. All rights reserved.
//

import MultipeerConnectivity
import GameKit
import UIKit

class UFOViewController: UIViewController, GameCenterManagerPlayerDelegate, GKGameCenterControllerDelegate, GKMatchmakerViewControllerDelegate {
    var gcManager: GameCenterManager?
    var peerPickerController: MCBrowserViewController?
    var currentSession: MCSession?
    var advertiserAssistant: MCAdvertiserAssistant?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(localUserAuthenticationChanged(_:)),
            name: .GKPlayerAuthenticationDidChangeNotificationName,
            object: nil)
        
        gcManager = GameCenterManager()
        gcManager?.playerDelegate = self
        gcManager?.authenticateLocalUser(self)
        gcManager?.populateAchievementCache()
        gcManager?.findAllActivity()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        gcManager?.playerDelegate = self
    }
    
    func playerDataLoaded(_ players: [GKPlayer]?, error: Error?) {
        if let error = error {
            print("An error occured during player lookup: \(error.localizedDescription)")
        } else {
            print("Players loaded: \(players ?? [])")
        }
    }
    
    @objc func localUserAuthenticationChanged(_ notif: Notification?) {
        if let object = notif?.object {
            print("Authenication Changed: \(object)")
        }
    }
    
    func processGameCenterAuthentication(_ error: Error?) {
        if let error = error {
            print("An error occured during authentication: \(error.localizedDescription)")
        }
    }
    
    func findProgrammaticMatch() {
        let request = GKMatchRequest()
        request.minPlayers = 2
        request.maxPlayers = 4


        GKMatchmaker.shared().findMatch(for: request, withCompletionHandler: { match, error in
            if let error = error {
                print("An error occurrred during finding a match: \(error.localizedDescription)")
            } else if let match = match {
                print("A match has been found: \(match)")
            }
        })
    }
    
    func findProgrammaticHostedMatch() {
        let request = GKMatchRequest()
        request.minPlayers = 2
        request.maxPlayers = 16
        
        GKMatchmaker.shared().findPlayers(forHostedRequest: request) { players, error in
            if let error = error {
                print("An error occurrred during finding a match: \(error.localizedDescription)")
            } else if let players = players {
                print("Players have been found for match: \(players)")
            }
        }
    }
    
    func addPlayer(to match: GKMatch?, with request: GKMatchRequest?) {
        if let match = match, let request = request {
            GKMatchmaker.shared().addPlayers(to: match, matchRequest: request, completionHandler: { error in
                if let error = error {
                    print("An error occurrred during adding a player to match: \(error.localizedDescription)")
                } else {
                    print("A player has been added to the match")
                }
            })
        }
    }
    
    func matchmakerViewControllerWasCancelled(_ viewController: GKMatchmakerViewController) {
        dismiss(animated: true, completion: nil)
    }
    
    func matchmakerViewController(_ viewController: GKMatchmakerViewController, didFailWithError error: Error) {
        dismiss(animated: true, completion: nil)
        
        let alert = UIAlertController.init(title: "", message: "An error occurred: \(error.localizedDescription)", preferredStyle: .alert)
        let defaultAction = UIAlertAction(title: "Dismiss", style: .default, handler: { action in
            self.dismiss(animated: true, completion: nil)
        })
        
        alert.addAction(defaultAction)
        present(alert, animated: true, completion: nil)
    }
    
    func matchmakerViewController(_ viewController: GKMatchmakerViewController, didFind match: GKMatch) {
        dismiss(animated: true, completion: nil)
    }
    
    func matchmakerViewController(_ viewController: GKMatchmakerViewController, didFindPlayers playerIDs: [String]) {
        dismiss(animated: true, completion: nil)
        print("Players: \(playerIDs)")
        //Begin Hosted Game
    }
    
    func playerActivity(_ activity: Int?, error: Error?) {
        if let error = error {
            print("An error occurred while querying player activity: \(error.localizedDescription)")
        } else {
            print("All recent player activity: \(activity ?? 0)")
        }
        
    }
    
    func playerActivityForGroup(_ activityDict: [AnyHashable : Any]?, error: Error?) {
        if let error = error {
            print("An error occurred while querying player activity: \(error.localizedDescription)")
        } else {
            if let activity = activityDict?["activity"],
               let playerGroup = activityDict?["playerGroup"] {
                print("All recent player activity: \(activity) For group: \(playerGroup)")
            }
        }
    }
    
    func friendsFinishedLoading(_ friends: [GKPlayer]?, error: Error?) {
        if let error = error {
            print("An error occured during friends list request: \(error.localizedDescription)")
        } else if let friends = friends {
            playerDataLoaded(friends, error: error)
        }
    }
    
    @IBAction func playButtonPressed() {
        let gameViewController = UFOGameViewController()
        gameViewController.gcManager = gcManager
        navigationController?.pushViewController(gameViewController, animated: true)
    }

    @IBAction func leaderboardButtonPressed() {
        let leaderboardController = GKGameCenterViewController.init(state: .default)
        leaderboardController.gameCenterDelegate = self
        present(leaderboardController, animated: true)
    }

    @IBAction func customLeaderboardButtonPressed() {
        let leaderboardViewController = UFOLeaderboardViewController()
        leaderboardViewController.gcManager = gcManager
        present(leaderboardViewController, animated: true)
    }
    
    func gameCenterViewControllerDidFinish(_ gameCenterViewController: GKGameCenterViewController) {
        dismiss(animated: true, completion: nil)
    }
    

    @IBAction func achievementButtonPressed() {
        var achievementViewController: GKGameCenterViewController? = nil
        if let state = GKGameCenterViewControllerState(rawValue: 1) {
            achievementViewController = GKGameCenterViewController(state: state)
        }
        achievementViewController?.gameCenterDelegate = self
        if let achievementViewController = achievementViewController {
            present(achievementViewController, animated: true)
        }
    }

    @IBAction func customAchievementButtonPressed() {
        let achievementViewController = UFOAchievementViewController()
        achievementViewController.gcManager = gcManager
        present(achievementViewController, animated: true)
    }

    @IBAction func multiplayerButtonPressed() {
        let request = GKMatchRequest()
        request.minPlayers = 2
        request.maxPlayers = 4
        
        guard let matchmakerViewController = GKMatchmakerViewController(matchRequest: request) else {
            print("There was an error creating the matchmaker view controller.")
            return
        }
        
        matchmakerViewController.matchmakerDelegate = self
        matchmakerViewController.isHosted = true
        
        present(matchmakerViewController, animated: true)
    }
    
    @IBAction func localMultiplayerPressed(_ sender: UIButton) {
        let request = GKMatchRequest()
        request.minPlayers = 2
        request.maxPlayers = 4

        let matchmakerViewController = GKMatchmakerViewController(matchRequest: request)
        matchmakerViewController?.matchmakerDelegate = self

        if let matchmakerViewController = matchmakerViewController {
            present(matchmakerViewController, animated: true)
        }
    }
}
